python/pecos/qeclib/steane/steane_class.py

```diff
@@ -97,115 +97,94 @@ def __init__(self, name: str, default_rus_limit: int = 3, ancillas: QReg | None
 
         self.default_rus_limit = default_rus_limit
 
-    def p(self, state: str, rus_limit: int | None = None):
+    def p(self, state: str, reject: Bit | None = None, rus_limit: int | None = None):
         """Prepare a logical qubit in a logical Pauli basis state."""
-        if rus_limit is None:
-            rus_limit = self.default_rus_limit
-
         match state:
             case "+X" | "X":
-                return self.px(rus_limit=rus_limit)
+                return self.px(reject=reject, rus_limit=rus_limit)
             case "-X":
-                return self.pnx(rus_limit=rus_limit)
+                return self.pnx(reject=reject, rus_limit=rus_limit)
             case "+Y" | "Y":
-                return self.py(rus_limit=rus_limit)
+                return self.py(reject=reject, rus_limit=rus_limit)
             case "-Y":
-                return self.pny(rus_limit=rus_limit)
+                return self.pny(reject=reject, rus_limit=rus_limit)
             case "+Z" | "Z":
-                return self.pz(rus_limit=rus_limit)
+                return self.pz(reject=reject, rus_limit=rus_limit)
             case "-Z":
-                return self.pnz(rus_limit=rus_limit)
+                return self.pnz(reject=reject, rus_limit=rus_limit)
             case _:
                 msg = f"State {state} is not implemented!"
                 raise NotImplementedError(msg)
 
-    def px(self, rus_limit: int | None = None):
+    def px(self, reject: Bit | None = None, rus_limit: int | None = None):
         """Prepare logical |+X>, a.k.a. |+>"""
-
-        if rus_limit is None:
-            rus_limit = self.default_rus_limit
-
         return PrepRUS(
             q=self.d,
             a=self.a[0],
             init=self.verify_prep[0],
-            limit=rus_limit,
+            limit=rus_limit or self.default_rus_limit,
             state="+X",
+            reject=reject or self.scratch[2],
```

Comment:
I think potentially moving the bit outside of the logical qubit object gets me bit nervous as a user might accidentally having different logical qubits override the same register. But fine if the value is constant. So `rus_limit` is probably okay. (Maybe worth adding a Const type.) 

Generally, I would like information generated by operations on logical qubits saved "privately" and when a user want it made public, to copy it to another place, e.g., see the pattern in the logical measurements.

Reply:
IMO this is fully on the user.  If I write a classical program, I get to decide when to set what variables, and how to use those variables throughout the program.  Here also I am allowing the user to set variables (bit registers) that user manages.  In principle, the user could even add things like conditional logic to the program based on the values of those variables.

I think it **does not** make sense to save all rejection bits to `Steane` class variables.  The Steane class only has (and, IMO, only should have) a "local" view of info/data attached to one logical qubit at the time of initialization.  It should not know global data such as (say) how many RUS protocols are in the program written by the user, which determine the size of a register that stores all RUS rejection bits.

Reply:
Sorry -- I am using strong language like "should" above for ease of communication.  I am ultimately just after data like RUS rejection bits and QEC/QED syndrome flags, so I'm happy to discuss how to best extract that info 🙂

Reply:
There might of been some confusion. I do not wish each Steane class to know the global information. It was more about privately protecting information locally for the Steane code and not let the users accidently override it by replacing the private register as I believe what you were code is proposing would do. Doing so could cause one register to be assigned to different independent codes.

Reply:
The option to pass a `reject` bit certainly lets the user override `Bit`s assigned to a `Steane` class -- which was also true before this PR.  Nothing prevents a user from using (say) `logical_qubit.flag_z[1]` somewhere in their `pecos.slr.Main` program.  Adding underscores to signal private data may be a good idea, but I would argue is outside the scope of this PR.

One point to note: the scratch bit that rejection data was stored to (and is still stored to by default) is not actually used for anything by the `Steane` class.  The class simply wrote rejection data to `self.scratch[2]`, and happily kept overriding that bit.  The same bit is also written over by various methods in `python/pecos/qeclib/steane/decoders/lookup.py`.  This PR simply lets the user store that data in a bit register of their choice.